### PR TITLE
Automatically use user token when sharing token is not sufficient for a request

### PR DESCRIPTION
**Pull request by MichaelBuessemeyer**
This PR improves the request retry functionality by trying using the token from the URL only upon the first try. If this results in a 403 (permissions not sufficient), the new code retries with the personal token of the user and if that succeeds, the URL token is removed and the previous error toast regarding a potentially expired token is hidden.

### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- create an annotation
- make it public but turn off editing by others
- copy the annotation link & open it, it should include a sharing token
- do some annotation actions & try saving
- the first saving try should fail & show an error toast
- Then another try should be done, which should succeed, remove the token from the URL, close the error toast and show another info toast about the retry.

### Issues:
- fixes [#7309](https://github.com/scalableminds/webknossos/issues/7309)

------
(Please delete unneeded items, merge only when none are left open)
- [x] Updated [changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)

## Release Notes

- **New Features**
  - Added metadata support for annotations on Trees and Segments.
  - Introduced a summary row in the time tracking overview.
  - Improved slider functionality and search capabilities for unnamed segments.
  - Enhanced loading speed for precomputed meshes.
  - Unified UI terminology for "Magnification."
  - Added support for remote OME-Zarr NGFF version 0.5 datasets.

- **Bug Fixes**
  - Resolved dataset upload failures with absolute paths.
  - Fixed issues with insufficient sharing tokens when saving annotations.
  - Corrected automatic group expansion in skeleton searches.
  - Addressed incorrect shape returns and sorting issues in zarr streaming.
  - Enabled annotation creation for public datasets of other organizations.

- **Improvements**
  - Enhanced error handling for user notifications and logging.
  - Adjusted error notification strategies for server requests.
  - Improved control flow for token retrieval based on availability.
  - Enhanced type safety in API request handling.

### CODE REVIEW

**Comment by MichaelBuessemeyer** [comment]
Sadly, the new change that auto removes a token from the URL if the personal token is more privileged does introduce quite some new cyclic dependencies.
Therefore, the auto removal is discarded for now and can be re-added later.

**Comment by MichaelBuessemeyer** [comment]
Here is the link about the follow up issue: https://github.com/scalableminds/webknossos/issues/8140

File: frontend/javascripts/admin/api/token.ts [hunk 1]
```diff
@@ -33,22 +35,48 @@ export function getSharingTokenFromUrlParameters(): string | null | undefined {
...
 
       // If three new tokens did not fix the 403, abort, otherwise we'll get into an endless loop here
       if (tries < 3) {
-        return doWithToken(fn, tries + 1);
+        // If using the url sharing token failed, we try the user specific token instead.
+        const result = await doWithToken(fn, tries + 1, false);
+        // Upon successful retry with own token, discard the url token.
+        if (useURLTokenIfAvailable) {
+          shouldUseURLToken = false;
+          Toast.info(
```
Comment (hotzenklotz) [inline frontend/javascripts/admin/api/token.ts:74]: I think this is a technical detail that we shouldn't bother users with. I would replace the `Toast` with a `console.log()` for debugging purposes but not actively show it in the UI.

Comment (MichaelBuessemeyer) [reply frontend/javascripts/admin/api/token.ts:74]: So just hiding the error toast is enough you think? My thought was to somewhat inform the user that saving / or some other request worked on second try.

Therefore, maybe a toast with a message like "Request succeeded on upon retry" or something similar?

Comment (MichaelBuessemeyer) [reply frontend/javascripts/admin/api/token.ts:74]: But yeah, I agree that the technical details are definitely too much in my toast message

**Comment by MichaelBuessemeyer** [comment]
We had a little internal discussion about how to handle the error toasts.

The solution we came up with is to disable the toast that shows the backend error why using the auto retry mechasim of `doWithToken` when annotation updates are sent to the server. The `save_saga` itself shows an error toast about the saving not working and retrying once the three auto retries of `doWithToken` failed.

In order to not let the backend error go completely silent, I changed the request handling in such a way that if no error toast is shown, the error is at least logged to the console. (Showing an error toast itself already automatically logs the error to the console. Thus, whenever a request error occurs, the error is logged to the console, it is irrelevant what the request `showErrorToast` config setting says.

@daniel-wer Maybe you also have an opinion on this. I am open for suggestions how to maybe improve this behaviour

**Comment by hotzenklotz** [comment]
@coderabbitai review

**Pull request merged**